src/proof/lfsc/lfsc_printer.cpp

```diff
@@ -80,91 +80,43 @@ void LfscPrinter::print(std::ostream& out,
   {
     // note that we must get all "component types" of a type, so that
     // e.g. U is printed as a sort declaration when we have type (Array U Int).
-    std::unordered_set<TypeNode> ctypes;
-    expr::getComponentTypes(st, ctypes);
-    for (const TypeNode& stc : ctypes)
+    ensureTypeDefinitionPrinted(preamble, st, sts, tupleArity);
+  }
+  // print datatype definitions for the above sorts
+  for (const TypeNode& stc : sts)
+  {
+    if (!stc.isDatatype() || stc.getKind() == PARAMETRIC_DATATYPE)
     {
-      if (sts.find(stc) != sts.end())
-      {
-        continue;
-      }
-      sts.insert(stc);
-      if (stc.isSort())
+      // skip the instance of a parametric datatype
+      continue;
+    }
+    const DType& dt = stc.getDType();
+    preamble << "; DATATYPE " << dt.getName() << std::endl;
+    NodeManager* nm = NodeManager::currentNM();
+    for (size_t i = 0, ncons = dt.getNumConstructors(); i < ncons; i++)
+    {
+      const DTypeConstructor& cons = dt[i];
+      std::string cname = d_tproc.getNameForUserNameOf(cons.getConstructor());
+      // for now, must print as node to ensure same policy for printing
+      // variable names. For instance, this means that cvc.X is printed as
+      // LFSC identifier |cvc.X| if X contains symbols legal in LFSC but not
+      // SMT-LIB. We should disable printing quote escapes in the smt2
```

Comment:
It is not clear whether “we should” is already implemented or not. If not, perhaps open an issue and mention it here?